File: epi_judge_cpp_solutions/tree_connect_leaves.cc
#include <memory>
#include <vector>

#include "binary_tree_node.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"

using std::unique_ptr;
using std::vector;

void AddLeavesLeftToRight(const unique_ptr<BinaryTreeNode<int>>&,
                          vector<const unique_ptr<BinaryTreeNode<int>>*>*);

vector<const unique_ptr<BinaryTreeNode<int>>*> CreateListOfLeaves(
    const unique_ptr<BinaryTreeNode<int>>& tree) {
  vector<const unique_ptr<BinaryTreeNode<int>>*> leaves;
  AddLeavesLeftToRight(tree, &leaves);
  return leaves;
}

void AddLeavesLeftToRight(
    const unique_ptr<BinaryTreeNode<int>>& tree,
    vector<const unique_ptr<BinaryTreeNode<int>>*>* leaves_ptr) {
  if (tree != nullptr) {
    if (tree->left == nullptr && tree->right == nullptr) {
      leaves_ptr->emplace_back(&tree);
    } else {
      AddLeavesLeftToRight(tree->left, leaves_ptr);
      AddLeavesLeftToRight(tree->right, leaves_ptr);
    }
  }
}

vector<int> CreateListOfLeavesWrapper(
    TimedExecutor& executor, const unique_ptr<BinaryTreeNode<int>>& tree) {
  auto result = executor.Run([&] { return CreateListOfLeaves(tree); });

  if (std::any_of(std::begin(result), std::end(result),
                  [](const unique_ptr<BinaryTreeNode<int>>*& x) {
                    return !x || !*x;
                  })) {
    throw TestFailure("Result list can't contain nullptr");
  }

  vector<int> extracted_result;
  for (const auto& x : result) {
    extracted_result.push_back(x->get()->data);
  }
  return extracted_result;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "tree"};
  return GenericTestMain(args, "tree_connect_leaves.cc",
                         "tree_connect_leaves.tsv", &CreateListOfLeavesWrapper,
                         DefaultComparator{}, param_names);
}

File: epi_judge_java_solutions/epi/TreeConnectLeaves.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;
import epi.test_framework.TimedExecutor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TreeConnectLeaves {

  public static List<BinaryTreeNode<Integer>>
  createListOfLeaves(BinaryTreeNode<Integer> tree) {

    List<BinaryTreeNode<Integer>> leaves = new ArrayList<>();
    addLeavesLeftToRight(tree, leaves);
    return leaves;
  }

  private static void
  addLeavesLeftToRight(BinaryTreeNode<Integer> tree,
                       List<BinaryTreeNode<Integer>> leaves) {
    if (tree != null) {
      if (tree.left == null && tree.right == null) {
        leaves.add(tree);
      } else {
        addLeavesLeftToRight(tree.left, leaves);
        addLeavesLeftToRight(tree.right, leaves);
      }
    }
  }

  @EpiTest(testDataFile = "tree_connect_leaves.tsv")
  public static List<Integer>
  createListOfLeavesWrapper(TimedExecutor executor,
                            BinaryTreeNode<Integer> tree) throws Exception {
    List<BinaryTreeNode<Integer>> result =
        executor.run(() -> createListOfLeaves(tree));

    if (result.stream().anyMatch(x -> x == null || x.data == null)) {
      throw new TestFailure("Result can't contain null");
    }

    List<Integer> extractedRes = new ArrayList<>();
    for (BinaryTreeNode<Integer> x : result) {
      extractedRes.add(x.data);
    }
    return extractedRes;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "TreeConnectLeaves.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
